Account for alignment changes when growing or shrinking

yanchith opened the pull request:
My riff on #143

Both `Bump::grow` and `Bump::shrink` now take in the new layout and check for alignment compatibility. `Bump::grow` falls back to `Bump::try_alloc_layout` if the alignment is not compatible, and `Bump::shrink` just returns `AllocError` in this case.

These above strategies are pretty much up to us. The Allocator trait says it is ok to error out, if it can't handle the new requirements. For bump allocation, I'd personally prefer shrinking to fail instead of allocating, if the alignment requirements changed.

Afaik, the collections in alloc don't change alignment during their life (unless transmuted), and while I can imagine legitimate code that would want to do this, the same code would probably write their own allocator.

Also added two tests, but `allocator_grow_layout_change` doesn't really test anything just yet. It is a bit tricky to figure out when grow falls back to `Bump::try_alloc_layout` without instrumenting Bump itself with counters. Ideas?

fitzgen commented:
> These above strategies are pretty much up to us. The Allocator trait says it is ok to error out, if it can't handle the new requirements. For bump allocation, I'd personally prefer shrinking to fail instead of allocating, if the alignment requirements changed.

Agreed, this sounds like the correct trade off to me.

fitzgen commented on tests/allocator_api.rs:
```diff
@@ -1,4 +1,4 @@
-#![feature(allocator_api)]
+#![feature(allocator_api, slice_ptr_get)]
```
Could you avoid the use of new nightly features here? It just makes things harder when we have to update unrelated tests because the feature has changed in some breaking way since the last PR was made. I know that the tests will be slightly more verbose, but it is worth it to avoid forcing churn upon unrelated PRs.

fitzgen commented on tests/allocator_api.rs:
```diff
@@ -114,3 +114,31 @@ fn allocator_grow_zeroed() {
     assert_eq!(unsafe { p.as_ref() }, [42, 42, 42, 42, 0, 0, 0, 0]);
 }
+
+#[test]
+fn allocator_grow_layout_change() {
+    let b = AllocatorDebug::new(Bump::with_capacity(1024));
+
+    let layout_align4 = Layout::from_size_align(256, 4).unwrap();
+    let layout_align16 = Layout::from_size_align(1024, 16).unwrap();
+
+    // Allocate a chunk of memory and attempt to grow it while increasing
+    // alignment requirements.
+    let p4 = b.allocate(layout_align4).unwrap().as_non_null_ptr();
+    let _p16 = unsafe { b.grow(p4, layout_align4, layout_align16).unwrap() };
+}
+
+#[test]
+fn allocator_shrink_layout_change() {
+    let b = AllocatorDebug::new(Bump::with_capacity(1024));
+
+    let layout_align4 = Layout::from_size_align(1024, 4).unwrap();
+    let layout_align16 = Layout::from_size_align(256, 16).unwrap();
+
+    // Allocate a chunk of memory and attempt to shrink it while increasing
+    // alignment requirements.
+    let p4 = b.allocate(layout_align4).unwrap().as_non_null_ptr();
+    let p16_res = unsafe { b.shrink(p4, layout_align4, layout_align16) };
+
+    assert_eq!(p16_res, Err(AllocError));
+}
```
Regarding being unsure whether the tests exercise the fallback-to-new-alloc paths, I think we could have much more confidence in exercising this code if we wrote targeted quickcheck tests instead of unit tests for these APIs.

I'm imagining that we would do something like:

```rust
quickcheck! {
    fn allocator_grow_layout_change(layouts: Vec<(usize, usize)>) -> () {
        let mut layouts = layouts.into_iter().map(|(size, align)| {
            let align = std::cmp::min(align, 64);
            let align = round_up_to_pow2(align);
            let size = std::cmp::min(size, 1024);
            let size = std::cmp::max(size, align);
            Layout::from_size_align(size, align).unwrap()
        }).collect::<Vec<_>>();
        layouts.sort_by_key(|l| (l.size(), l.align()));
        // TODO: allocate the first layout in the bump and then grow
        // to each of the next layouts
    }
}
```

where `round_up_to_pow2` is basically this helper but without checked arithmetic (safe because we are clamping the input just above): https://cs.github.com/bytecodealliance/wasmtime/blob/b438617e1238bec01ebc804c0087e4849df0dc87/crates/runtime/src/externref.rs?q=round_up#L291

Exercising shrinking would be the exact same except that we would invert the comparator so that the layouts were ordered form largest to smallest.

Does that make sense?

yanchith replied:
Yes, this makes sense. Do we want to assert something in these quickchecks, or is it enough that they don't panic for now?

fitzgen replied:
I think we can start with just asserting that the grown/shrunk pointer has the requested alignment.

fitzgen reviewed (commented):
Thanks! This is looking great. Just some suggestions for how to better exercise these paths in tests.

yanchith commented:
Sooo, I added those quickchecks, and actually found what looks like a real issue in `Bump::shrink`: applying the size delta to the pointer could actually produce pointers with insufficient alignment (because the size could actually be any "unround" integer once it is GTE the alignment). I fixed this by aligning the delta, so doing arithmetic with it doesn't change alignment of the pointer. We reclaim less space this way, but the new pointer stays correctly aligned.

This is all very new to me, so please double-check 😂

fitzgen reviewed (commented):
Thanks! I think the `shrink` code was assuming that the size would always be a multiple of the alignment but it is just greater-than-or-equal. Good find!

Just one nitpick to fix below and then we can merge this.

fitzgen reviewed (approved):
Thanks! And thanks for sticking with this PR throughout a couple rounds of feedback!
